### ISSUE TITLE
Lounge crashes when user logs in from different system (LDAP enabled)

### ISSUE THREAD

**Issue by thisisdarshan** [thelounge/lounge#666]
Lounge crashes after the user logs in from a different system.  For the context, I have enabled LDAP for authentication.  It could be related the case (notice the first time user logged in with "Xyz" and the second time the user logged in with "xyz").  Here is the trace log -

```
2016-09-20 23:55:40 [INFO] User 'abc' loaded
2016-09-20 23:55:40 [INFO] User 'bcd' loaded
2016-09-20 23:59:59 [INFO] User 'Xyz' loaded
2016-09-21 16:31:56 [INFO] User 'xyz' loaded
Missing error handler on `socket`.
TypeError: Cannot read property 'replace' of undefined
    at ldapAuth (/bnc/codebase/src/server.js:264:19)
    at Socket.auth (/bnc/codebase/src/server.js:321:4)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)
    at Socket.onevent (/bnc/codebase/node_modules/socket.io/lib/socket.js:335:8)
    at Socket.onpacket (/bnc/codebase/node_modules/socket.io/lib/socket.js:295:12)
    at Client.ondecoded (/bnc/codebase/node_modules/socket.io/lib/client.js:193:14)
    at Decoder.Emitter.emit (/bnc/codebase/node_modules/socket.io/node_modules/socket.io-parser/node_modules/component-emitter/index.js:134:20)
    at Decoder.add (/bnc/codebase/node_modules/socket.io/node_modules/socket.io-parser/index.js:247:12)
    at Client.ondata (/bnc/codebase/node_modules/socket.io/lib/client.js:175:18)
events.js:141
      throw er; // Unhandled 'error' event
      ^

Error: write EPIPE
    at exports._errnoException (util.js:907:11)
    at WriteWrap.afterWrite (net.js:785:14)

npm ERR! Linux 3.10.0-327.28.3.el7.x86_64
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "start"
npm ERR! node v4.5.0
npm ERR! npm  v2.15.9
npm ERR! code ELIFECYCLE
npm ERR! thelounge@2.0.0-rc.1 start: `node index`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the thelounge@2.0.0-rc.1 start script 'node index'.
npm ERR! This is most likely a problem with the thelounge package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node index
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs thelounge
npm ERR! Or if that isn't available, you can get their info via:
npm ERR! 
npm ERR!     npm owner ls thelounge
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /bnc/codebase/npm-debug.log
```

**Comment by astorije** [comment]
@thisisdarshan, is this LDAP-only? If so, do you think you would be able to fix it?
If this is LDAP-related, maybe @lindskogen can chime in as well :-)

**Comment by gramakri** [comment]
We hit a similar backtrace on the Cloudron app. Cloudron app also has LDAP enabled. https://git.cloudron.io/cloudron/thelounge-app/issues/1 has a similar backtrace.

**Comment by gramakri** [comment]
@astorije Atleast one crash is easily reproducible. Simply press "Sign in" button with the username empty and LDAP enabled. The code crashes as below:
```
16:45:36 [main] /home/cloudron/src/server.js:311
16:45:36 [main] 	var userDN = user.replace(/([,\\/#+<>;"= ])/g, "\\$1");
16:45:36 [main] 	                 ^
16:45:36 [main] 
16:45:36 [main] TypeError: Cannot read property 'replace' of undefined
16:45:36 [main]     at ldapAuth (/home/cloudron/src/server.js:311:19)
16:45:36 [main]     at Socket.auth (/home/cloudron/src/server.js:378:4)
16:45:36 [main]     at emitOne (events.js:77:13)
16:45:36 [main]     at Socket.emit (events.js:169:7)
16:45:36 [main]     at /home/cloudron/node_modules/socket.io/lib/socket.js:503:12
16:45:36 [main]     at nextTickCallbackWith0Args (node.js:436:9)
16:45:36 [main]     at process._tickCallback (node.js:365:13)
```

The reason is that in `client/js/lounge.js` form submit callback, the `user` property is not set at all when it's empty. This causes data.user on the server side code to be undefined and thus causes a crash.

**Comment by astorije** [comment]
@gramakri problem with this is that none of the @thelounge/maintainers use LDAP nor have enough knowledge of it to maintain it properly.
We need help from the community to keep it working across evolution of the software. Would you or anyone else using The Lounge with LDAP be able to help us with that? No hard commitments needed, just be around to help with LDAP issues or making sure every now and then (like, new releases) that it works okay, and help fixing it if not.

Thanks!

**Comment by gramakri** [comment]
@astorije Sure, feel free to tag me for any LDAP related issues. We test it on lounge+ldap on Cloudron regularly, so any LDAP issue interests us.

As for this crash, I will send a PR.

**Comment by astorije** [comment]
That's awesome, thanks @gramakri!
I'll assign you to any outstanding stuff regarding LDAP, feel free to not take all of them, but at least you will be aware.

@thelounge/maintainers, do we want a GitHub Project for LDAP stuff? Just curious.

**Comment by astorije** [comment]
Actually, I can't assign you because you need to be issue author or team member, so I'll just Cc you for now.

**Comment by xPaw** [comment]
@eliemichel Can you check if this is not a problem in your PR?

**Comment by eliemichel** [comment]
I've added a `if (!user) { callback(false); }` from the beginning. As far as I remember, it was for this kind of issue, although I was not aware of the present bug report.

I think that when the user session is remembered, the front-end tries to authenticate as soon as the page loads, but when the token or so has expired it results in an auth with an undefined user, which I decided to just ignore. I know tha ldapAuth function but not fully the conditions in which it is called.
